### PR TITLE
[BugFix] Force cte resue when plan has non-deterministic function

### LINKED ISSUE #31647: Incorrect query results, possibly due to incorrect inlining of UUID function
### Steps to reproduce the behavior (Required)
```
unzip [bug.zip](https://github.com/StarRocks/starrocks/files/12701386/bug.zip)
cd bug-repro
./bug.sh
```
### Expected behavior (Required)
If you diff correct.sql and bug.sql, you can see that there's a small change that should not affect the results. I expect the to SQLs to produce equivalent results:

```
Sleeping for 1 minute
Creating database
Running Correct.sql
JoinColumnResolver_Column_ValueTable_000
9
1
Running Bug.sql
9
1
Removing docker container
starrocks-test
starrocks-test
```

### Real behavior (Required)
They produce different results. My guess is that the UUID is getting inlined somewhere and causing it to produce different UUIDs for the same CTE and thus we are not joining as we expect to.

```
Sleeping for 1 minute
Creating database
Running Correct.sql
JoinColumnResolver_Column_ValueTable_000
9
1
Running Bug.sql
Removing docker container
starrocks-test
starrocks-test
```

### StarRocks version (Required)
starrocks/allin1-ubuntu:3.1.0-rc01

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #31647

Currently, when deciding whether to use CTEs (Common Table Expressions), we do not take into account whether non-deterministic functions are present within the operators. This can result in unexpected outcomes.

Add a rule to enforce cte when non-deterministic functions are present within the operator.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 24 / 94 (25.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java|22|92|23.91%|[58, 59, 63, 71, 72, 74, 75, 76, 78, 79, 88, 89, 90, 91, 93, 94, 95, 96, 97, 100, 101, 102, 104, 105, 109, 110, 111, 112, 113, 115, 116, 117, 121, 122, 124, 125, 126, 127, 129, 130, 131, 133, 134, 140, 148, 157, 164, 165, 166, 167, 169, 174, 175, 176, 178, 183, 184, 185, 187, 192, 193, 194, 195, 196, 198, 199, 204, 205, 206, 208]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                    hasNonDeterministicFunc(joinOperator.getOnPredicate())) {
+                return true;
+            }
+            return visitChildren(optExpression);
+        }
+
+        @Override
+        public Boolean visitLogicalAggregate(OptExpression optExpression, Void context) {
+            LogicalAggregationOperator operator = (LogicalAggregationOperator) optExpression.getOp();
+            if (checkAggCall(operator.getAggregations())) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java:176]: Having Predicate？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                }
+            }
+            return false;
+        }
+
+        private boolean checkOptExpression(OptExpression optExpression) {
+            Operator operator = optExpression.getOp();
+            if (operator.getProjection() != null && checkProject(operator.getProjection())) {
+                return true;
+            }
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java:136]: Predicate can be pushed down to each operator. check predicate process may like projection.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java]: fixed.

**Review by packy92 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 70 / 93 (75.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java|68|91|74.73%|[72, 75, 76, 78, 92, 108, 116, 117, 119, 120, 121, 122, 124, 125, 126, 128, 129, 136, 164, 174, 183, 192, 213]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 70 / 93 (75.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java|68|91|74.73%|[72, 75, 76, 78, 92, 108, 116, 117, 119, 120, 121, 122, 124, 125, 126, 128, 129, 136, 164, 174, 183, 192, 213]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 3]
```diff
@@ -349,6 +350,7 @@ private OptExpression logicalRuleRewrite(ConnectContext connectContext,
         if (cteContext.needOptimizeCTE()) {
             cteContext.reset();
             ruleRewriteOnlyOnce(tree, rootTaskContext, RuleSetType.COLLECT_CTE);
+            ruleRewriteOnlyOnce(tree, rootTaskContext, new ForceCTEReuseRule());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:353]: should execute it after `PUSH_DOWN_PREDICATE` (line:361)?

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 70 / 93 (75.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java|68|91|74.73%|[72, 75, 76, 78, 92, 108, 116, 117, 119, 120, 121, 122, 124, 125, 126, 128, 129, 136, 164, 174, 183, 192, 213]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 70 / 93 (75.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ForceCTEReuseRule.java|68|91|74.73%|[72, 75, 76, 78, 92, 108, 116, 117, 119, 120, 121, 122, 124, 125, 126, 128, 129, 136, 164, 174, 183, 192, 213]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**